File: features/auth/express/sequelize/files/src/models/refresh-token.ts
import { DataTypes, Model } from 'sequelize';
import { sequelize } from '../db/client.js';

export class RefreshToken extends Model {
  declare id: string;
  declare user_id: string;
  declare session_id: string;
  declare token_hash: string;
  declare ip_address: string | null;
  declare user_agent: string | null;
  declare expires_at: Date;
  declare revoked_at: Date | null;
  declare rotated_to: string | null;
  declare replay_detected_at: Date | null;
  declare created_at: Date;
  declare updated_at: Date;
}

RefreshToken.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    user_id: {
      type: DataTypes.UUID,
      allowNull: false,
      references: { model: 'users', key: 'id' },
      onDelete: 'CASCADE',
    },
    session_id: {
      type: DataTypes.UUID,
      allowNull: false,
    },
    token_hash: {
      type: DataTypes.STRING(64),
      allowNull: false,
      unique: true,
    },
    ip_address: {
      type: DataTypes.STRING(64),
      allowNull: true,
    },
    user_agent: {
      type: DataTypes.TEXT,
      allowNull: true,
    },
    expires_at: {
      type: DataTypes.DATE,
      allowNull: false,
    },
    revoked_at: {
      type: DataTypes.DATE,
      allowNull: true,
    },
    rotated_to: {
      type: DataTypes.UUID,
      allowNull: true,
    },
    replay_detected_at: {
      type: DataTypes.DATE,
      allowNull: true,
    },
  },
  {
    sequelize,
    modelName: 'RefreshToken',
    tableName: 'refresh_tokens',
    underscored: true,
    indexes: [{ fields: ['user_id'] }, { fields: ['session_id'] }],
  },
);

File: features/auth/rust/files/src/auth/models/refresh_token.rs
use chrono::{DateTime, Utc};
use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Serialize, Deserialize)]
#[sea_orm(table_name = "refresh_tokens")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false, column_type = "Uuid")]
    pub id: Uuid,
    pub user_id: Uuid,
    pub session_id: Uuid,
    #[sea_orm(unique)]
    pub token_hash: String,
    pub ip_address: Option<String>,
    pub user_agent: Option<String>,
    pub expires_at: DateTime<Utc>,
    pub revoked_at: Option<DateTime<Utc>>,
    pub rotated_to: Option<Uuid>,
    pub replay_detected_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {}

impl ActiveModelBehavior for ActiveModel {}

impl Model {
    pub fn active(
        user_id: Uuid,
        session_id: Uuid,
        token_hash: String,
        ip_address: Option<String>,
        user_agent: Option<String>,
        expires_at: DateTime<Utc>,
    ) -> ActiveModel {
        let now = Utc::now();
        ActiveModel {
            id: sea_orm::Set(Uuid::new_v4()),
            user_id: sea_orm::Set(user_id),
            session_id: sea_orm::Set(session_id),
            token_hash: sea_orm::Set(token_hash),
            ip_address: sea_orm::Set(ip_address),
            user_agent: sea_orm::Set(user_agent),
            expires_at: sea_orm::Set(expires_at),
            revoked_at: sea_orm::Set(None),
            rotated_to: sea_orm::Set(None),
            replay_detected_at: sea_orm::Set(None),
            created_at: sea_orm::Set(now),
            updated_at: sea_orm::Set(now),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use sea_orm::ActiveValue;

    #[test]
    fn new_populates_required_fields() {
        let uid = Uuid::new_v4();
        let sid = Uuid::new_v4();
        let exp = Utc::now();
        let am = Model::active(uid, sid, "hash".into(), Some("1.2.3.4".into()), None, exp);
        assert!(matches!(am.user_id, ActiveValue::Set(v) if v == uid));
        assert!(matches!(am.session_id, ActiveValue::Set(v) if v == sid));
        assert!(matches!(am.revoked_at, ActiveValue::Set(None)));
        assert!(matches!(am.rotated_to, ActiveValue::Set(None)));
    }
}
